Improve performance of weeder when type-class-roots = false is set.

pranaysashank opened the pull request:
For analysing evidence uses we collect evidence uses in,

```haskell
requestedEvidence :: Map Declaration (Set Name)
```
In `analyseEvidenceUses`, we loop over all the names in all the sets of the map, to construct dependency graph after calling `getEvidenceTree` on the name. However, these names in sets across different declarations are duplicated a lot. In one example in a repo at work, we have `16961625` names in which only `200330` are unique. So now, we instead pre-construct an evidence trees map `Map Name [Declaration]` for all the unique name and perform a lookup in this map to construct the graph.

In a private repo, the times before this change and after

```
❯ find . -name '*.hie'  | wc -l
   1097

❯ time result/bin/weeder # weeder from master

real    5m53.707s
user    5m50.350s
sys     0m2.206s

 ❯ time result/bin/weeder # weeder from this branch

real    0m34.008s
user    0m31.716s
sys     0m2.196s
```

I have a failing test in `./test/Spec/ModuleRoot`, but this is failing on master as well.

ocharles commented:
Wow, what an incredible speed up! I'll take a look! @ryndubei could you also have a look if you have time?

ryndubei commented on src/Weeder.hs:
```diff
@@ -730,22 +731,38 @@ requestEvidence n d = do
-   in star d evBindSiteDecls
+-- |Construct a map from the evidence to their declarations following the
+-- given evidence uses back to their instance bindings
+evidenceDeclarationMap :: RefMap TypeIndex -> Set Name -> Map Name [Declaration]
+evidenceDeclarationMap rf names = Set.foldl' go mempty names
```
Using `Map.fromSet` instead of a fold should be a bit better asymptotically, and would make it explicit that the resulting map preserves the contents of `names`.

Actually - what would be even better is to change the type signature to just 
```haskell
evidenceDeclarations :: RefMap TypeIndex -> Name -> [Declaration]
```
and let `analyseEvidenceUses` build the map. Then we could use `Map.!` as mentioned in a lower comment without having to look at the implementation of `evidenceDeclarations` to justify it.

pranaysashank replied:
That's how I had it initially! But then moved it around as that mostly preserved existing organization of the two functions

ryndubei reviewed (commented):
Thanks! I don't have a large enough project on hand at the moment that I can quickly test the performance with, but the changes look good to me. I only have a few minor, mostly-style-related comments.

pranaysashank commented on src/Weeder.hs:
```diff
@@ -730,22 +731,38 @@ requestEvidence n d = do
 
--- | Follow evidence uses listed under 'requestedEvidence' back to their 
+    go m n =
+      if Map.member n m
+        then m
+        else
```
Ah! True

pranaysashank commented:
@ryndubei I addressed all of your comments

one minor thing I have noticed is there seems to be a slight performance loss with the usage of `foldMap` & `Map.fromSet`. Core shows that `fromSet` doesn't get inlined and `foldMap` generates recursive code with `(++)` operators instead as opposed to when we hand roll `(concat (map (declMap Map.!) (Set.toList v)))` & use `Set.foldl'`. The same is the case with `fromSet` (turns out even foldl' has the same O(n) complexity). A quick look at [`fromSet`](https://hackage.haskell.org/package/containers-0.7/docs/src/Data.Map.Strict.Internal.html#fromSet) shows it doesn't have the inline pragma or the usual haskell trick where the recursion is extracted to a separate go function.

Since the performance impact only seems to be at most 1-2 seconds, I think we can prefer code clarity

pranaysashank commented:
also, why don't we link the executable with `-rtsopts`? it's useful to change some parameters to figure out how the performance characteristics change or limit the memory used with `+RTS -M`

ocharles commented:
Before...

```
[root@internal:~]# time nix-store -r /nix/store/sc5pfnj0la...-weeder.drv --builders '' --check
real	6m1.956s
user	6m5.595s
sys	0m2.792s
```

After

```
[root@internal:~]# time nix-store -r /nix/store/kip37...-weeder.drv --builders '' --check
real	3m5.243s
user	3m11.471s
sys	0m3.609s
```

Hard to argue with that!

Thanks for you hard work, @pranaysashank!
